File: JavaScript/0116. Populating Next Right Pointers in Each Node.js
// Given a binary tree
//
// struct TreeLinkNode {
//   TreeLinkNode *left;
//   TreeLinkNode *right;
//   TreeLinkNode *next;
// }
//
// Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.
//
//   Initially, all next pointers are set to NULL.
//
// Note:
//
//   You may only use constant extra space.
//   Recursive approach is fine, implicit stack space does not count as extra space for this problem.
//   You may assume that it is a perfect binary tree (ie, all leaves are at the same level, and every parent has two children).
//
// Example:
//
// Given the following perfect binary tree,
//
//     1
//    /  \
//   2    3
//  / \  / \
// 4  5  6  7
//
// After calling your function, the tree should look like:
//
//     1 -> NULL
//    /  \
//   2 -> 3 -> NULL
//  / \  / \
// 4->5->6->7 -> NULL

/**
 * // Definition for a Node.
 * function Node(val,left,right,next) {
 *    this.val = val;
 *    this.left = left;
 *    this.right = right;
 *    this.next = next;
 * };
 */
/**
 * @param {Node} root
 * @return {Node}
 */

// 1)
// Time O(n)
// Space O(1)
const connect1 = (root) => {
  if (root == null || root.left == null) return root;

  root.left.next = root.right; // connect left -> right
  root.right.next = root.next ? root.next.left : null; // connect right -> next's left

  connect(root.left);
  connect(root.right);

  return root;
};

// 2) BFS - level-order traversal
// Similar
// 102. Binary Tree Level Order Traversal
// 116. Populating Next Right Pointers in Each Node
//
// Time O(n)
const connect = (root) => {
  if (root == null) return null;

  let q = [root];

  while (q.length) {
    const nodes = [...q];
    q = [];
    while (nodes.length) {
      const node = nodes.shift();
      node.next = nodes.length ? nodes[0] : null;

      if (node.left) q.push(node.left);
      if (node.right) q.push(node.right);
    }
  }
  return root;
};

File: Python/0116. Populating Next Right Pointers in Each Node.py
# You are given a perfect binary tree where all leaves are on the same level, and every parent has two children. The binary tree has the following definition:
#
# struct Node {
#   int val;
#   Node *left;
#   Node *right;
#   Node *next;
# }
#
# Populate each next pointer to point to its next right node. If there is no next right node, the next pointer should be set to NULL.
# Initially, all next pointers are set to NULL.
#
# Example 1:
#
# Input: root = [1,2,3,4,5,6,7]
# Output: [1,#,2,3,#,4,5,6,7,#]
# Explanation: Given the above perfect binary tree (Figure A), your function should populate each next pointer to point to its next right node, just like in Figure B. The serialized output is in level order as connected by the next pointers, with '#' signifying the end of each level.
#
# Example 2:
#
# Input: root = []
# Output: []
#
# Constraints:
#
# The number of nodes in the tree is in the range [0, 212 - 1].
# -1000 <= Node.val <= 1000
#
# Follow-up:
#
# You may only use constant extra space.
# The recursive approach is fine. You may assume implicit stack space does not count as extra space for this problem.


"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

# 1)
# Time O(n)
# Space O(1)
class Solution:
    def connect(self, root: "Optional[Node]") -> "Optional[Node]":
        if not root:
            return None
        if root.left:
            root.left.next = root.right
        if root.right:
            root.right.next = root.next.left if root.next else None
        self.connect(root.left)
        self.connect(root.right)
        return root


# 2) BFS - level-order traversal
# Similar
# 102. Binary Tree Level Order Traversal
# 116. Populating Next Right Pointers in Each Node
#
# Time O(n)
class Solution:
    def connect(self, root: "Optional[Node]") -> "Optional[Node]":
        if not root:
            return None

        q = [root]

        while q:
            nodes = q.copy()
            q = []
            while nodes:
                node = nodes.pop(0)
                node.next = nodes[0] if nodes else None
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
        return root
